### DISCUSSION TITLE
Support for nested Menus

### DISCUSSION THREAD

**Discussion by mrodrigues95** [tailwindlabs/headlessui#669] [Ideas]
Hi,

This is somewhat related to #152 

Are there any plans for supporting nested menus? My current use case is that when you select a menu item I need to re-render the list of items in the menu and show them without closing it. This is pretty straight forward but the issue is the menu closes on the first click.

I'm aware I can handle the open state manually but then I need to deal with handling outside clicks, etc. which is already managed by Headless UI.

**Comment by RobinMalfait** [comment]
Hey! Thank you for your suggestion!
Much appreciated! 🙏

This is something we want to support, but I can't give you an ETA on this.
We will first focus on other components, but we will get back to this.

**Comment by benknight** [comment]
+1

**Comment by benknight** [comment]
I think this would be rather easy to implement manually if you just provided a way to open a menu programmatically. As of now it can only be opened by clicking on Menu.Button. Just provide an `open` prop on the Menu component, then we can just render another Menu inside Menu.Item and pass the `active` render prop to the Menu's  `open` prop.  Wouldn't that work?

**Comment by ShivamJoker** [comment]
any updates on this ?

**Reply by brianwachira** [reply]
It's mid 2022, also curious about the updates

**Reply by steven-wood** [reply]
^

**Reply by pkvov** [reply]
^

**Comment by vitorcamachoo** [comment]
Is there any way to do this?

**Comment by gmwill934** [comment]
Any update on this? I'm trying to implement a sidebar multi level menu? Are there any other alternatives?

**Comment by bonesoul** [comment]
I'm also looking for this.

**Comment by vixducis** [comment]
I think this is now possible since 1.7.4. In this version, a close method has been exposed on the Menu which can be used to manually close the menu. I put a quick example together in a stackblitz: https://stackblitz.com/edit/react-ts-iuxwoh

It's pretty crude, but you should get the idea.

**Reply by JasonPaff** [reply]
Good stuff @vixducis, I just have to get the keyboard handling working correctly and its exactly what I needed, thanks!

**Reply by bonesoul** [reply]
Can you show an example on the keyboard stuff?

**Reply by knownasilya** [reply]
Also stuck on keyboard, nesting a menu inside `Menu.Item` seems to break activating the `Menu.Button` with the keyboard (it just closes the main menu)

**Comment by ahmadaldabouqii** [comment]
?`

**Comment by techwithanirudh** [comment]
I put a quick example together in a stackblitz based on @vixducis's example:  https://stackblitz.com/edit/react-ts-2gynxv?file=App.tsx

**Reply by zacktagnan** [reply]
@techwithanirudh ... In the absence of this multi-level functionality in the Menu component of this package, would it be possible to have a version of this React example adapted for Vue? Thanks.

**Comment by Zhengqbbb** [comment]
Nesting a `Menu` inside `Menu.Item` seems to break activating the `Menu.Button` with the keyboard (it **closes** the main menu that make a bad a11y !!!)

**Comment by PavelZhuravlev** [comment]
this feature would be awesome improvement for menu functionality!

**Comment by IRediTOTO** [comment]
I would love have one :D

**Comment by anthonyboutinov** [comment]
This is horrible! I spent hours, trying everything I could think of to hack it to work, but the keyboard navigation breaks if you nest menus—you cannot focus the trigger button using keyboard navigation if it sits inside a parent menu.

We just have to abandon healessui in our team… I loved it for the easy documentation but alas.😔

The examples from other people here are not of idiomatic submenus, they're primitive workarounds that break proper UX.

**Reply by spaceemotion** [reply]
We are using a mix of headlessUI + radix-vue for this reason. After a ton of tries, HUI wasn't able to produce any nested menus, modals, etc.

**Reply by knownasilya** [reply]
We rolled our own menus using floatingui, it was a lot of work.

**Comment by briavicenti** [comment]
@RobinMalfait Any chance we can get an update here? Looks like a lot of folks (myself included) are needing to move away from HeadlessUI because this feature is missing.

**Comment by arshad-yaseen** [comment]
Please +1

**Comment by Fabioni** [comment]
it's 5 years now ...

**Comment by y4qq** [comment]
+1

**Comment by Yukinosuke-Takada** [comment]
Here's the workaround for v2.X that I used for my project.

I discovered that Menu, Listbox etc works inside Popover. So I'm using Popover for the parent menu. So Simply instead of `Menu>MenuButton+MenuItems>MenuItem>Menu`, `Popover>PopoverButton+PopoverPanel>Menu`.

Here's some example: https://stackblitz.com/edit/react-ts-2mc6rkdl?file=App.tsx 

I know this isn't the intended use and i might have broken some a11y (but visually & keyboard navigation seems to work fine).
